### ISSUE TITLE
Run-TestsInNavContainer fails with AAD authentication

### ISSUE THREAD

**Issue by richardrobberse** [microsoft/navcontainerhelper#521]
Hi Freddy,

Did you ever manage to run tests in a container with AAD authentication?
The scenario is a bit weird, but in our release pipeline we'd like to have the possibility to run our tests once again to make sure that apps are not breaking each other.

If I call the function without a credential object, I get the error that I need to specify a credential object for NavUserPassword authentication, which makes sense because the 'ClientServicesCredentialType' on the service tier is NavUserPassword in case of AAD auth.

If I call the function with the correct credentials, I get the generic client session error:
CommunicationError : Error while copying content to a stream.
ClientSession in Error
at AwaitState, C:\ProgramData\NavContainerHelper\Extensions\navserver\PsTestTool-1\ClientContext.ps1: line 98
at InvokeInteraction, C:\ProgramData\NavContainerHelper\Extensions\navserver\PsTestTool-1\ClientContext.ps1: line 111
at InvokeInteractionAndCatchForm, C:\ProgramData\NavContainerHelper\Extensions\navserver\PsTestTool-1\ClientContext.ps1: line 120
at OpenForm, C:\ProgramData\NavContainerHelper\Extensions\navserver\PsTestTool-1\ClientContext.ps1: line 135
at Run-Tests, C:\ProgramData\NavContainerHelper\Extensions\navserver\PsTestTool-1\PsTestFunctions.ps1: line 136
at <ScriptBlock>, <No file>: line 37
ClientSession in Error
At C:\Program Files\WindowsPowerShell\Modules\navcontainerhelper\0.6.2.6\ContainerHandling\Invoke-ScriptInNavContainer.ps1:30 char:13
+             Invoke-Command -Session $session -ScriptBlock $scriptbloc ...
+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (ClientSession in Error:String) [], RuntimeException
    + FullyQualifiedErrorId : ClientSession in Error

I've also tried to create a new NavUserPassword user and pass that one in as the credential object, this results in the same generic client session error.
In both cases, this is the error that ends up in the event log of the container:
_You do not have access to Dynamics 365 Business Central. Verify that AUTOMATEDTESTS is set up as a valid Dynamics 365 Business Central user._

I can imagine that this is simply not supported, but I just want to be sure :)

**Comment by freddydk** [comment]
It is on my todo list to fix, but today it is not supported.

**Comment by richardrobberse** [comment]
I see, thanks for the update.

**Comment by freddydk** [comment]
I looked at this today - and I have checked in a solution in the dev branch, which supports that you specify an AccessToken to Get-Tests and Run-Tests.

It is the responsibility of your code to get the access token - then you can run the tests.

Note that getting the access token is different if your AAD requires MFA - but the following code works for accounts created by demos.microsoft.com.

Note that the c:\demo\settings.ps1 contains the clientid and clientsecret if you have setup an Azure VM with AAD auth.

```
. "C:\DEMO\navcontainerhelper-dev\NavContainerHelper.ps1"
. "C:\DEMO\settings.ps1"

$client = New-Object System.Net.Http.HttpClient

$authority = "https://login.microsoftonline.com/Common/"
$resource = $ssoAdAppId
$clientId = $ssoAdAppId
$clientSecret = $ssoAdAppKeyValue

$content = New-Object “System.Collections.Generic.Dictionary[[System.String],[System.String]]”
$content.Add("resource", $resource)
$content.Add("client_id", $clientId)
$content.Add("grant_type", "password")
$content.Add("username", $username)
$content.Add("password", $password)
$content.Add("scope", "openid")
$content.Add("client_secret", $clientSecret)

$form = New-Object System.Net.Http.FormUrlEncodedContent -ArgumentList $content

$result = $client.PostAsync("$authority/oauth2/token", $form).Result
$resultJson = $result.Content.ReadAsStringAsync().Result | ConvertFrom-Json
$accessToken = $resultJson.access_token

$credential = New-Object pscredential 'admin', (ConvertTo-SecureString -String 'Pepsimax4ever' -AsPlainText -Force)
Get-TestsFromBCContainer -containerName navserver -credential $credential -accessToken $accessToken
Run-TestsInNavContainer -containerName navserver -credential $credential -accessToken $accessToken -detailed
```

If this useful?

**Comment by freddydk** [comment]
Implemented in 0.6.2.9

**Comment by richardrobberse** [comment]
Thanks for the fix Freddy!
